core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -604,14 +604,14 @@ class AdminManager(val config: KafkaConfig,
         case OpType.APPEND => {
           if (!listType(alterConfigOp.configEntry().name(), configKeys))
             throw new InvalidRequestException(s"Config value append is not allowed for config key: ${alterConfigOp.configEntry().name()}")
-          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name(), "").split(",").toList.filter(_.trim.length > 0)
           val newValueList = oldValueList ::: alterConfigOp.configEntry().value().split(",").toList
           configProps.setProperty(alterConfigOp.configEntry().name(), newValueList.mkString(","))
         }
         case OpType.SUBTRACT => {
           if (!listType(alterConfigOp.configEntry().name(), configKeys))
             throw new InvalidRequestException(s"Config value subtract is not allowed for config key: ${alterConfigOp.configEntry().name()}")
-          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name(), "").split(",").toList.filter(_.trim.length > 0)
           val newValueList = oldValueList.diff(alterConfigOp.configEntry().value().split(",").toList)
```

Comment:
An open question is if we subtract everything, should that remove the property or just leave it as an empty string?

Reply:
I think, we should remove the config.

Reply:
After spending several hours arguing with the test, I realized that a "removal" (delete) reverts the property to the default. So if we have a property whose list value is "foo, bar" and we SUBTRACT "foo, bar" ... is the property now set to "" or to <default>? (Especially if default IS, say, "foo")

Reply:
I think, we should fallback to broker/topic defaults after removal of any overrides. currently, we should the same for  OpType.DELETE operation. We should also follow the same for OpType. SUBTRACT?

Reply:
I think it makes more sense to end up with an empty list if you SUBTRACT each list element. That's how lists typically work. We have DELETE if the desired behavior is to actually remove the config.

Reply:
My main concern is setting empty list to configs like `cleanup.policy` may lead to unintended errors. 

cc @rajinisivaram

Reply:
May be worth looking at all the broker and topic configs of type LIST and seeing whether empty list makes sense in most cases. If it doesn't. perhaps throwing an exception may be better than deleting quietly since you may end up with a default value that you didn't intend for the config.

Reply:
Thanks @rajinisivaram ! -- a discussion with others led to the realization that empty string is the wrong answer in the general case, I have updated to use the stringified version of the config default.

Reply:
I have just pushed a version which stays with the "subtract from the list even if the list goes to empty", and will momentarily push a version which does the "config removal/revert to default when subtract yields an empty list". I defer to the Kafka committers on this.

Reply:
I think it makes sense to apply append/subtract on the config default
value rather than "", if the config is previously not set. 

The only issue with config removal/revert to default is we can not set empty list to config value.

Reply:
We could always force users to use the `SET` operation in this case, rather than `SUBTRACT`.
However, on thinking about it, the current code will let you SUBTRACT to an empty list. So in the interest of keeping this a bug fix (not breaking existing code/changing behavior, just preventing the NPE), maybe we should not include my last pushed commit and allow SUBTRACT to generate an empty list. This is rope, but it is existing rope.

Reply:
makes sense. we can revert last commit and handle separately in another JIRA.

Can we please update the javadoc with current behaviour of `OpType.SET`, 
`OpType.DELETE`, `OpType.APPEND` and `OpType.SUBTRACT` operations?
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/Admin.java#L394

Reply:
FYI, validated in the Javadoc for AlterConfigOp.OpType --

![image](https://user-images.githubusercontent.com/42848865/76537037-685c9480-643a-11ea-842b-e0376580bf61.png)